clang/lib/Sema/SemaDecl.cpp

```diff
@@ -8236,11 +8236,14 @@ void Sema::CheckShadow(NamedDecl *D, NamedDecl *ShadowedDecl,
   DeclContext *NewDC = D->getDeclContext();
 
   if (FieldDecl *FD = dyn_cast<FieldDecl>(ShadowedDecl)) {
-    // Fields are not shadowed by variables in C++ static methods.
-    if (CXXMethodDecl *MD = dyn_cast<CXXMethodDecl>(NewDC))
+    if (CXXMethodDecl *MD = dyn_cast<CXXMethodDecl>(NewDC)) {
+      // Fields are not shadowed by variables in C++ static methods.
       if (MD->isStatic())
         return;
 
+      if (!MD->getParent()->isLambda() && MD->isExplicitObjectMemberFunction())
+        return;
```

Comment:
Do we have a test for the lambda case? E.g.

```c++
struct S {
  int x;
  void f() {
    [&](this auto& self, int x) {}; // warning
  }
};
```

Reply:
Maybe this also warrants a comment, because at first it wasn’t immediately obvious to me why we’d want to ignore lambda call operators here.

Reply:
```
```

Reply:
@Sirraide Thanks for the feedback. I've added an additional test. 

> Maybe this also warrants a comment, because at first it wasn’t immediately obvious to me why we’d want to ignore lambda call operators here.

Thanks for bringing that to my attention. In such cases

```cpp
struct C {
  int b = 5;
  int foo() {
    return [a = b]() {
      return [=, b = a]() {
        return b;
      }();
    }();
  }
};
```

depending exclusively on `isExplicitObjectMemberFunction` isn’t safe, as it could lead to assertion failures during type resolution

https://github.com/llvm/llvm-project/blob/1a590870b6b3452934ecc245e01957fdab48909c/clang/lib/AST/Decl.cpp#L3699-L3702

There’s an option to restrict the null type or to validate the parent scope. I've been trying to find a suitable utility to do this, but I haven’t had any luck. If there’s one available, I’d appreciate it if you could point me to it. thanks

Reply:
> Thanks for bringing that to my attention. In such cases
> 
> ```c++
> struct C {
>   int b = 5;
>   int foo() {
>     return [a = b]() {
>       return [=, b = a]() {
>         return b;
>       }();
>     }();
>   }
> };
> ```
> 
> depending exclusively on `isExplicitObjectMemberFunction` isn’t safe, as it could lead to assertion failures during type resolution

Er, can you elaborate what the problem here is maybe?

Reply:
@Sirraide apologies for the missing details in my previous comment. When the checker encounters a _lambda call operator_ (which uses `CXXMethodDecl`) with an incomplete type, it fails because it can't retrieve parameters from a `null` type.

Reply:
It makes sense that that would fail, but I’m still a bit confused as to how we end up w/ a `CXXMethodDecl` whose type (after desugaring) isn’t a `FunctionProtoType`. If that *does* happen somehow, I should hope that `isInvalidDecl()` also returns `true`, in which case you’d check for that first.

Reply:
We might want a similar function that does more or less the opposite in that case (i.e. an `isLambdaCallWithExplicitObjectParameter`)?

CC @cor3ntin

Reply:
@Sirraide I'm fine with adding it. I'm unsure how widely it will be used, however having such a utility seems reasonable. @cor3ntin what do you think?

Reply:
@Sirraide it does exist., however, it doesn't handle the case described above

https://github.com/llvm/llvm-project/blob/08e7609692af3cb84da510deac70eeb02cbceb6d/clang/include/clang/AST/ASTLambda.h#L38-L41